File: core/src/main/java/com/zp4rker/iab/api/Explorer.java
package com.zp4rker.iab.api;

import com.j256.ormlite.field.DatabaseField;
import com.j256.ormlite.table.DatabaseTable;
import com.zp4rker.iab.IABCore;
import org.bukkit.entity.Player;

import java.sql.SQLException;
import java.util.UUID;

@DatabaseTable(tableName = "explorers")
public class Explorer {
    @DatabaseField(id = true)
    private UUID uuid;
    @DatabaseField(unique = true, canBeNull = false)
    private String name;
    @DatabaseField(foreign = true, foreignAutoRefresh = true)
    private Spaceship spaceship;

    @DatabaseField
    private int tripCount = 0;
    @DatabaseField
    private float flightTime = 0;
    @DatabaseField
    private float distanceTravelled = 0;
    @DatabaseField
    private int planetsVisited = 0;

    public Explorer(Player player, String name) {
        this.uuid = player.getUniqueId();
        this.name = name;
    }

    public Explorer() {
    }

    public Explorer(UUID uuid, String name, int tripCount, float flightTime, float distanceTravelled, int planetsVisited) {
        this.uuid = uuid;
        this.name = name;
        this.tripCount = tripCount;
        this.flightTime = flightTime;
        this.distanceTravelled = distanceTravelled;
        this.planetsVisited = planetsVisited;
    }

    public UUID getUUID() {
        return uuid;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Spaceship getSpaceship() {
        return spaceship;
    }

    public void setSpaceship(Spaceship spaceship) {
        this.spaceship = spaceship;
    }

    public int getTripCount() {
        return tripCount;
    }

    public void incrementTripCount() {
        tripCount++;
    }

    public float getFlightTime() {
        return flightTime;
    }

    public void addFlightTime(float flightTime) {
        this.flightTime += flightTime;
    }

    public float getDistanceTravelled() {
        return distanceTravelled;
    }

    public void addDistanceTravelled(float distanceTravelled) {
        this.distanceTravelled += distanceTravelled;
    }

    public int getPlanetsVisited() {
        return planetsVisited;
    }

    public void incrementPlanetsVisited() {
        planetsVisited++;
    }

    public boolean save() {
        try {
            IABCore.DB_MANAGER.saveExplorer(this);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    public boolean delete() {
        try {
            IABCore.DB_MANAGER.deleteExplorer(this);
            return true;
        } catch (SQLException e) {
            return false;
        }
    }

    public boolean inDatabase() {
        return find(uuid) != null;
    }

    public static Explorer find(Player player) {
        return find(player.getUniqueId());
    }

    public static Explorer find(UUID uuid) {
        try {
            return IABCore.DB_MANAGER.findExplorer(uuid);
        } catch (SQLException e) {
            return null;
        }
    }
}

File: core/src/main/kotlin/com/zp4rker/iab/api/Explorer.kt
package com.zp4rker.iab.api

import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.table.DatabaseTable
import com.zp4rker.iab.DB_MANAGER
import org.bukkit.entity.Player
import java.sql.SQLException
import java.util.UUID

@DatabaseTable(tableName = "explorers")
class Explorer() {
    @DatabaseField(id = true)
    var uuid: UUID = UUID.randomUUID()
    @DatabaseField(unique = true)
    var name: String = ""
    @DatabaseField(foreign = true, foreignAutoRefresh = true)
    var spaceship: Spaceship = Spaceship()

    @DatabaseField
    var tripCount: Int = 0
    @DatabaseField
    var flightTime: Float = 0f
    @DatabaseField
    var distanceTravelled: Float = 0f

    constructor(player: Player, name: String) : this() {
        this.uuid = player.uniqueId
        this.name = name
    }

    fun save(): Boolean {
        try {
            DB_MANAGER?.saveExplorer(this) ?: throw SQLException("Database manager not setup!")
            return true
        } catch (e: SQLException) {
            return false
        }
    }

    fun delete(): Boolean {
        try {
            DB_MANAGER?.deleteExplorer(this) ?: throw SQLException("Database manager not setup!")
            return true
        } catch (e: SQLException) {
            return false
        }
    }

    fun inDatabase() = find(uuid) != null

    companion object {
        fun find(player: Player) = find(player.uniqueId)

        fun find(uuid: UUID): Explorer? {
            return try {
                DB_MANAGER?.findExplorer(uuid) ?: throw SQLException("Database manager not setup!")
            } catch (e: SQLException) {
                null
            }
        }
    }
}
